### ISSUE TITLE
SC2153 confuses local variables from other functions

### ISSUE THREAD

**Issue by dfong** [koalaman/shellcheck#518]
this is in shellcheck-0.4.1.  i did not see it in shellcheck-0.3.5 which i had been using before.
example:
```
#! /bin/bash
# test script for shellcheck SC2153

# defines local $myvar
F()
{
        local myvar=1
        echo "in F: myvar=$myvar"
}

# uses global $MYVAR.  note that $myvar should be out of scope
G()
{
        echo "in G: MYVAR=$MYVAR"
        echo "in G: myvar=$myvar"
}

# ----- start of mainline code
F
G
```
when i run shellcheck on the above script, i get:
```
In t.sh line 14:
	echo "in G: MYVAR=$MYVAR"
                          ^-- SC2153: Possible misspelling: MYVAR may not be assigned, but myvar is.
```
note that at line 14, "myvar" from F is out of scope.

**Comment by kurahaupo** [comment]
This is not provable in the general case, as shellcheck would need to prove
that F does not call G. (Note that scoping is dynamic, not lexical.)

**Comment by dfong** [comment]
point noted.  however, the message says "myvar is" assigned.  in order for that assignment to be pertinent, shellcheck would need to prove (or assume) that F *does* call G.  even with dynamic scoping, it seems like a stretch.

**Comment by koalaman** [comment]
Yes, this is an issue. Shellcheck basically just finds that a variable by this similar name is at some point assigned, while the variable you reference is not.

It shouldn't be implying that it's currently in scope.

As a side note, being in scope would not be strict requirement due to inheritance and accidental subshell.

**Comment by romanek-adam** [comment]
I was hit by this issue too.

Basically I have a variable which is passed to a script through the environment and there's a function which defines a local variable with a similar name (lowercase).

    $ cat test.sh
    #!/bin/bash

    set -e

    func() {
      local foo_bar=${1}
      echo "local: ${foo_bar}"
    }

    echo "FOO_BAR: ${FOO_BAR}"

    func "${FOO_BAR}"

Lets assume that the script is run like this:

    FOO_BAR=X ./test.sh

Now when using shellcheck 0.3.7:

    $ shellcheck test.sh 

    In test.sh line 10:
    echo "FOO_BAR: ${FOO_BAR}"
                   ^-- SC2153: Possible misspelling: FOO_BAR may not be assigned, but foo_bar is.

I thought that using `[ -v FOO_BAR ]` to check whether a variable is defined would solve the problem:

    $ cat test2.sh 
    #!/bin/bash

    set -e

    func() {
      local foo_bar=${1}
      echo "local: ${foo_bar}"
    }

    [ -v FOO_BAR ] || (echo "FOO_BAR is not defined"; exit 1)

    echo "FOO_BAR: ${FOO_BAR}"

    func "${FOO_BAR}"

but it didn't:

    $ shellcheck test2.sh 

    In test2.sh line 10:
    [ -v FOO_BAR ] || (echo "FOO_BAR is not defined"; exit 1)
         ^-- SC2153: Possible misspelling: FOO_BAR may not be assigned, but foo_bar is.
         ^-- SC2086: Double quote to prevent globbing and word splitting.

What would you recommend to overcome this issue?

**Comment by kurahaupo** [comment]
On Fri, 29 Jul 2016, Adam Romanek wrote:
> I thought that using `[ -v FOO_BAR ]` to check whether a variable is defined would solve the problem:

Shellcheck does not check programmatic flow, it only does static analysis.

Using the variable as `${FOO_BAR-default_foooo_baaaa}` should silence
"undefined" warnings though.

**Comment by romanek-adam** [comment]
OK I understand shellcheck limitations. Nevertheless in my case there's no default for the variable, it needs to be passed or the script should fail. This is why I use `-v FOO_BAR`.

**Comment by kurahaupo** [comment]
On Fri, 29 Jul 2016, Adam Romanek wrote:
> OK I understand Shellcheck limitations.

That's a tad disingenuous; deciding whether some code gets executed is not
generally soluble -- it's equivalent to The Halting Problem.

The Halting Problem has itself been formally proven: it is impossible to
prove or disprove whether every program will halt or not. (For some simple
programs yes, but for any given proving method, some programs will be
undecidable.)

In the face of exported functions, it is basically impossible for
Shellcheck to know exactly what your program will do, without actually
running it (which would rather defeat the point).

So Shellcheck takes the approach of issuing warnings unless it can prove
that the warning is not necessary. Failing that, if the human can prove it,
they're welcome to add a "disable" annotation.

> Nevertheless in my case there's no default for the variable, it needs to
> be passed or the script should fail. This is why I use `-v FOO_BAR`.

In that case use something like:

```${SomeVar?'You must export SomeVar into this program'}```

**Comment by koalaman** [comment]
Shellcheck currently just assumed each statement in the program is run in series, which is a good oversimplification (though it's worked relatively well in practice for simple scripts). 

It's a bug that some basic data flow analysis would solve in this particular case, and that's on the radar..

**Comment by ChillerDragon** [comment]
I now prefix my local variable with a underscore to not get it confused with the ENV var used in another function. But would be nice if we could add scope support :)

**Comment by mkolesnik** [comment]
This is still relevant on version 0.8.0 :(

BTW it seems that if the local variable (and the env variable) don't have underscores this doesnt happen

So this will pass fine:
```bash
function bar() {
    local foo="$1"
    echo "${foo}"
}

echo "${FOO}"
```

While this fails:
```bash
function bar() {
    local foo_v="$1"
    echo "${foo_v}"
}

echo "${FOO_V}"
```

**Comment by kurahaupo** [comment]
aside: shouldn't ShellCheck warn when both `function` _and_ `()` are used together, like:
```lang=sh
function bar() {
```
? (Use one or the other but not both.)

**Comment by koalaman** [comment]
> This is still relevant on version 0.8.0 :(

Yes :( I wish I could take a month off my day job and get this done, but that has not been an option so far.

> BTW it seems that if the local variable (and the env variable) don't have underscores this doesnt happen

It's actually the length. ShellCheck's heuristic is "the variable might be misspelled if it's 4+ characters long and 1 char apart, or 7+ chars long and up to 2 chars apart". The point of this is to not trigger on `$bar` vs `$baz`, and do trigger on `$referer` and `$referrer`. For the purpose of this check, two strings that only differ in case is considered 1 char apart, and thus it somewhat unintentionally ends up only triggering on 4+ character strings.

>aside: shouldn't ShellCheck warn when both function and () are used together

It's valid in Bash so there's no warning against it, but if you do it in Ksh you get https://github.com/koalaman/shellcheck/wiki/SC2111

**Comment by kurahaupo** [comment]
> It's valid in Bash so there's no warning against it

I know it's allowed, but it seems to me like a gratuitous incompatibility that should be discouraged as it's a difference without any purpose, so I'd like to see [SC2111](../wiki/SC2111) included in the warnings for Bash as well, at least when in "pedantic" mode.

One reason I dislike this difference is because it means that we can't simply translate
```bash
foo() ( echo bar )
```
to
```bash
function foo ( echo bar )
```
To be fair, we can't yet write that in Ksh either, where the grammar says <tt>function <i>NAME</i> { <i>LIST</i> ; }</tt> (ie, a brace-block), but Ksh could relax that rule without breaking existing code, whereas Bash would have yet another messy "look 2 tokens ahead" to maintain compatibility both ways.
